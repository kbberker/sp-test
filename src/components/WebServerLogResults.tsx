interface WebServerLogResultsProps {
  rawLogData: string[];
}

interface PageVisitDetails {
  page: string;
  totalPageVisits: number;
  uniqueVisits: number;
}

const createPageVisitDetails = (log: string[], logIndex: number, allPageVisits: string[][]) => {
  const totalPageVisits = allPageVisits.filter((listItem) => log[0] === listItem[0]);
  const ipAddresses = totalPageVisits.map((pageVisit) => pageVisit[1]);
  const uniqueVisits = ipAddresses.filter(
    (ipAddress, index) => ipAddresses.indexOf(ipAddress) === index,
  );
  return {
    page: log[0],
    totalPageVisits: totalPageVisits.length,
    uniqueVisits: uniqueVisits.length,
  };
};

const WebServerLogResults = ({ rawLogData }: WebServerLogResultsProps): JSX.Element => {
  const formattedListOfSiteVisits = rawLogData.map((log) => log.split(' '));
  const pagesVisited = formattedListOfSiteVisits.map((siteVisit) => siteVisit[0]);

  const listOfPageVisitDetails: PageVisitDetails[] = formattedListOfSiteVisits.map(createPageVisitDetails);

  const listOfUniquePageVisitDetails = listOfPageVisitDetails.reduce<PageVisitDetails[]>(
    (list, siteVisit, currentIndex) => {
      if (pagesVisited.indexOf(siteVisit.page) === currentIndex && siteVisit.page !== '') {
        list.push(siteVisit);
      }
      return list;
    }, [],
  );

  const sortedByTotalVisits = [...listOfUniquePageVisitDetails].sort((a, b) => b.totalPageVisits - a.totalPageVisits);
  const sortedByUniqueVisits = [...listOfUniquePageVisitDetails].sort((a, b) => b.uniqueVisits - a.uniqueVisits);

  return (
    <div>
      <h2>Total Visits</h2>
      <ul>
        {listOfUniquePageVisitDetails.length > 0
          ? sortedByTotalVisits.map((pageVisitDetails) => (
            <li key={pageVisitDetails.page}>
              {`${pageVisitDetails.page} - ${pageVisitDetails.totalPageVisits} visits`}
            </li>
          )) : <p>No data to display</p>}
      </ul>
      <h2>Unique Visits</h2>
      <ul>
        {listOfUniquePageVisitDetails.length > 0
          ? sortedByUniqueVisits.map((pageVisitDetails) => (
            <li key={pageVisitDetails.page}>
              {`${pageVisitDetails.page} - ${pageVisitDetails.uniqueVisits} unique views`}
            </li>
          )) : <p>No data to display</p>}
      </ul>
    </div>
  );
};

export default WebServerLogResults;
export { createPageVisitDetails };
