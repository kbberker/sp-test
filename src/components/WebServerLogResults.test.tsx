import React from 'react';
import { render } from '@testing-library/react';
import WebServerLogResults, { createPageVisitDetails } from './WebServerLogResults';

const mockData = [
  '/help_page/1 126.318.035.038',
  '/contact 184.123.665.067',
  '/home 184.123.665.067',
  '/about/2 444.701.448.104',
  '/help_page/1 929.398.951.889',
  '/index 444.701.448.104',
  '/help_page/1 722.247.931.582',
  '/about 061.945.150.735',
  '/help_page/1 646.865.545.408',
  '/home 235.313.352.950',
  '/contact 184.123.665.067',
  '/help_page/1 543.910.244.929',
];

const mockFormattedData = [
  ['/help_page/1', '126.318.035.038'],
  ['/contact', '184.123.665.067'],
  ['/home 184.123.665.067'],
  ['/about/2', '444.701.448.104'],
  ['/help_page/1', '929.398.951.889'],
  ['/index 444.701.448.104'],
  ['/help_page/1', '722.247.931.582'],
  ['/about 061.945.150.735'],
  ['/help_page/1', '646.865.545.408'],
  ['/home', '235.313.352.950'],
  ['/contact', '184.123.665.067'],
  ['/help_page/1', '126.318.035.038'],
];

test('WebServerLogResults snapshot is correct', () => {
  const page = render(<WebServerLogResults rawLogData={mockData} />);
  expect(page).toMatchSnapshot();
});

test('createPageVisitDetails correctly creates pageVisitDetails', () => {
  expect(createPageVisitDetails(['/help_page/1', '126.318.035.038'], 0, mockFormattedData))
    .toStrictEqual({ page: '/help_page/1', totalPageVisits: 5, uniqueVisits: 4 });
});
