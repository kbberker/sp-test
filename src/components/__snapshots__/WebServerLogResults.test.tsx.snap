// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WebServerLogResults snapshot is correct 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div>
        <h2>
          Total Visits
        </h2>
        <ul>
          <li>
            /help_page/1 - 5 visits
          </li>
          <li>
            /contact - 2 visits
          </li>
          <li>
            /home - 2 visits
          </li>
          <li>
            /about/2 - 1 visits
          </li>
          <li>
            /index - 1 visits
          </li>
          <li>
            /about - 1 visits
          </li>
        </ul>
        <h2>
          Unique Visits
        </h2>
        <ul>
          <li>
            /help_page/1 - 5 unique views
          </li>
          <li>
            /home - 2 unique views
          </li>
          <li>
            /contact - 1 unique views
          </li>
          <li>
            /about/2 - 1 unique views
          </li>
          <li>
            /index - 1 unique views
          </li>
          <li>
            /about - 1 unique views
          </li>
        </ul>
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      <h2>
        Total Visits
      </h2>
      <ul>
        <li>
          /help_page/1 - 5 visits
        </li>
        <li>
          /contact - 2 visits
        </li>
        <li>
          /home - 2 visits
        </li>
        <li>
          /about/2 - 1 visits
        </li>
        <li>
          /index - 1 visits
        </li>
        <li>
          /about - 1 visits
        </li>
      </ul>
      <h2>
        Unique Visits
      </h2>
      <ul>
        <li>
          /help_page/1 - 5 unique views
        </li>
        <li>
          /home - 2 unique views
        </li>
        <li>
          /contact - 1 unique views
        </li>
        <li>
          /about/2 - 1 unique views
        </li>
        <li>
          /index - 1 unique views
        </li>
        <li>
          /about - 1 unique views
        </li>
      </ul>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
